packages/dds/tree/src/feature-libraries/sequence-change-family/compose.ts

```diff
@@ -0,0 +1,246 @@
+    splitMarkOnInput,
+    splitMarkOnOutput,
+    Transposed as T,
+} from "../../changeset";
+import { clone, fail, StackyIterator } from "../../util";
+import { SequenceChangeset } from "./sequenceChangeset";
+
+/**
+ * Composes a sequence of changesets into a single changeset.
+ * @param changes - The changesets to be applied.
+ * Parts of the input may be reused in the output, but the input is not mutated.
+ * Each changeset in the list is assumed to be applicable after the previous one.
+ * @returns A changeset that is equivalent to applying each of the given `changes` in order.
+ *
+ * WARNING! This implementation is incomplete:
+ * - Tombstone information is ignored.
+ * - Support for moves is not implemented.
+ * - Support for slices is not implemented.
+ */
+export function compose(changes: SequenceChangeset[]): SequenceChangeset {
+    if (changes.length === 1) {
+        return changes[0];
+    }
+    let composedFieldMarks: T.FieldMarks = {};
+    for (const change of changes) {
+        composedFieldMarks = composeFieldMarks(composedFieldMarks, change.marks);
+    }
+    return {
+        marks: composedFieldMarks,
+    };
+}
+
+function composeFieldMarks(baseFieldMarks: T.FieldMarks, newFieldMarks: T.FieldMarks): T.FieldMarks {
+    const composed: T.FieldMarks = {};
+    for (const key of Object.keys(newFieldMarks)) {
+        const composedMarkList = composeMarkLists(baseFieldMarks[key] ?? [], newFieldMarks[key]);
+        if (composedMarkList.length > 0) {
+            composed[key] = composedMarkList;
+        }
+    }
+    for (const key of Object.keys(baseFieldMarks)) {
+        if (!(key in newFieldMarks)) {
+            composed[key] = baseFieldMarks[key];
+        }
+    }
+    return composed;
+}
+
+function composeMarkLists(
+    baseMarkList: T.MarkList,
+    newMarkList: T.MarkList,
+): T.MarkList {
+    const factory = new MarkListFactory();
+    const baseIter = new StackyIterator(baseMarkList);
+    const newIter = new StackyIterator(newMarkList);
+    for (let newMark of newIter) {
+        let baseMark: T.Mark | undefined = baseIter.pop();
+        if (baseMark === undefined) {
+            // We have reached an region of the field that the base change does not affect.
+            // We therefore adopt the new mark as is.
+            factory.push(clone(newMark));
+        } else if (isAttachGroup(newMark)) {
+            // Content that is being attached by the new changeset cannot interact with base changes.
+            // Note that attach marks from different changesets can only target the same gap if they are concurrent.
+            // If two concurrent changesets A and B were targeting the same gap, the changeset that is sequenced later
+            // would bet rebased over the first, resulting in a Skip mark being introduced in the later changeset.
+            // Since compose only deals with changes that are rebased in this way,
+            // we don't have to deal with tie-breaking issues.
```

Comment:
I think it's easier to understand if we say less.

```suggestion
            // Note that attach marks from different changesets can only target the same gap if they are concurrent.
            // This method assumes that `newMarkList` is based on `baseMarkList`, so they are not concurrent.